stateproof/signer.go

```diff
@@ -83,83 +80,145 @@ func (spw *Worker) nextStateProofRound(latest basics.Round) basics.Round {
 		return
 	}
 
 	// Only sign blocks that are a multiple of StateProofInterval.
-	if hdr.Round%basics.Round(proto.StateProofInterval) != 0 {
+	if round%basics.Round(proto.StateProofInterval) != 0 {
+		return
+	}
+
+	if votersCommitment.IsEmpty() {
+		// No voter commitment, perhaps because state proofs were
+		// just enabled.
 		return
 	}
 
-	keys := spw.accts.StateProofKeys(hdr.Round)
+	keys := spw.accts.StateProofKeys(round)
 	if len(keys) == 0 {
 		// No keys, nothing to do.
 		return
 	}
 
-	// votersRound is the round containing the merkle root commitment
-	// for the voters that are going to sign this block.
-	votersRound := hdr.Round.SubSaturate(basics.Round(proto.StateProofInterval))
-	votersHdr, err := spw.ledger.BlockHdr(votersRound)
+	stateProofMessage, err := spw.getStateProofMessage(round, proto)
 	if err != nil {
-		spw.log.Warnf("spw.signBlock(%d): BlockHdr(%d): %v", hdr.Round, votersRound, err)
+		spw.log.Warnf("spw.signStateProof(%d): getStateProofMessage: %v", round, err)
 		return
 	}
 
-	if votersHdr.StateProofTracking[protocol.StateProofBasic].StateProofVotersCommitment.IsEmpty() {
-		// No voter commitment, perhaps because state proofs were
-		// just enabled.
-		return
+	spw.signStateProofMessage(stateProofMessage, round, keys)
+}
+
+func (spw *Worker) getProtoVoters(round basics.Round) (*config.ConsensusParams, crypto.GenericDigest, error) {
+	// throughout this function, we assume that the protocol version
+	// (specifically for our purposes, the state proof interval) is identical in the last attested round -
+	// which should be the round parameter in this case - and in the voters round.
+
+	signedHeader, err := spw.ledger.BlockHdr(round)
+	if err != nil {
+		return spw.getProtoVotersFallback(round)
 	}
 
-	sigs := make([]sigFromAddr, 0, len(keys))
+	proto := config.Consensus[signedHeader.CurrentProtocol]
+	votersRound := round.SubSaturate(basics.Round(proto.StateProofInterval))
+	votersHdr, err := spw.ledger.BlockHdr(votersRound)
 
-	stateproofMessage, err := GenerateStateProofMessage(spw.ledger, uint64(votersHdr.Round), hdr)
 	if err != nil {
-		spw.log.Warnf("spw.signBlock(%d): GenerateStateProofMessage: %v", hdr.Round, err)
-		return
+		return spw.getProtoVotersFallback(round)
+	}
+
+	return &proto, votersHdr.StateProofTracking[protocol.StateProofBasic].StateProofVotersCommitment, nil
+}
+
+func (spw *Worker) getProtoVotersFallback(round basics.Round) (*config.ConsensusParams, crypto.GenericDigest, error) {
+	verificationContext, err := spw.ledger.StateProofVerificationContext(round)
+	if err != nil {
+		return nil, crypto.GenericDigest{}, err
+	}
+
+	proto := config.Consensus[verificationContext.Version]
+	return &proto, verificationContext.VotersCommitment, nil
+}
+
+func (spw *Worker) getStateProofMessage(round basics.Round, proto *config.ConsensusParams) (*stateproofmsg.Message, error) {
+	dbBuilder, err := spw.loadBuilderFromDB(round)
+	if err == nil {
+		return &dbBuilder.Message, nil
```

Comment:
I think that the `builder` structure remains in memory although we only need the message. Maybe it is safer to return the message by value (so the builder can be disposed.)